File: pkg/bugnav/viewer/src/algorithms/distbug.rs
//! DistBug (Kamon & Rivlin, 1997). Bug2 with a range sensor — leave the wall
//! when `d_leave - F ≤ d_min - STEP`, where `F` is the free distance along
//! the ray to goal, as seen by the sensor.

use std::collections::HashSet;

use crate::algorithms::bug_common::{
    DIRS, VISION_R_SQ, WallFollowState, WallStepOutcome, dir_to_goal, dist_sq, neighbour,
    wall_follow_step,
};
use crate::grid::Grid;
use crate::pathfinder::{Pathfinder, Snapshot, StepStatus};

const STEP: f64 = 1.0;

#[derive(Clone, Copy, Debug, PartialEq, Eq)]
enum Mode {
    MotionToGoal,
    Follow,
}

pub struct DistBug {
    grid_w: i32,
    grid_h: i32,
    walls: Vec<bool>,
    goal: (i32, i32),
    pos: (i32, i32),
    mode: Mode,
    wf: WallFollowState,
    hit_point: (i32, i32),
    d_min: f64,
    follow_steps: u32,
    follow_visited: HashSet<((i32, i32), (i32, i32), bool)>,
    steps: u32,
    snap: Snapshot,
    status: StepStatus,
}

fn dist(a: (i32, i32), b: (i32, i32)) -> f64 {
    (dist_sq(a, b) as f64).sqrt()
}

pub fn build(grid: &Grid, start: (i32, i32), goal: (i32, i32)) -> Box<dyn Pathfinder> {
    let mut snap = Snapshot {
        current: start,
        path: vec![start],
        ..Snapshot::default()
    };
    snap.visited.insert(start);
    Box::new(DistBug {
        grid_w: grid.w,
        grid_h: grid.h,
        walls: grid.walls.clone(),
        goal,
        pos: start,
        mode: Mode::MotionToGoal,
        wf: WallFollowState {
            pos: start,
            current_obstacle: start,
            obstacle_on_right: true,
        },
        hit_point: start,
        d_min: dist(start, goal),
        follow_steps: 0,
        follow_visited: HashSet::new(),
        steps: 0,
        snap,
        status: StepStatus::Running,
    })
}

impl DistBug {
    fn passable(&self, x: i32, y: i32) -> bool {
        if x < 0 || y < 0 || x >= self.grid_w || y >= self.grid_h {
            return false;
        }
        !self.walls[(y * self.grid_w + x) as usize]
    }

    fn move_to(&mut self, new_pos: (i32, i32)) {
        self.pos = new_pos;
        self.snap.current = new_pos;
        self.snap.visited.insert(new_pos);
        self.snap.path.push(new_pos);
    }

    fn step_wall_follow(&mut self) -> WallStepOutcome {
        let (w, h, walls) = (self.grid_w, self.grid_h, &self.walls);
        let passable =
            |x: i32, y: i32| x >= 0 && y >= 0 && x < w && y < h && !walls[(y * w + x) as usize];
        let on_map = |x: i32, y: i32| x >= 0 && y >= 0 && x < w && y < h;
        let outcome = wall_follow_step(&mut self.wf, passable, on_map);
        if outcome == WallStepOutcome::Moved {
            self.move_to(self.wf.pos);
        }
        outcome
    }

    fn enter_follow(&mut self, blocked_dir: usize) {
        self.mode = Mode::Follow;
        self.wf = WallFollowState {
            pos: self.pos,
            current_obstacle: neighbour(self.pos, blocked_dir),
            obstacle_on_right: true,
        };
        self.hit_point = self.pos;
        self.follow_steps = 0;
        self.follow_visited.clear();
        self.follow_visited.insert((
            self.wf.pos,
            self.wf.current_obstacle,
            self.wf.obstacle_on_right,
        ));
    }

    /// Free distance along the king-direction ray toward the goal, measured
    /// in cells traversed (Euclidean distance to last passable cell). Must
    /// match the direction motion-to-goal will actually use — otherwise the
    /// leave condition can fire while motion is still blocked, causing an
    /// infinite Follow↔MotionToGoal oscillation.
    fn free_distance_to_goal(&self) -> f64 {
        let d = dir_to_goal(self.pos, self.goal);
        let (dx, dy) = DIRS[d];
        let mut p = self.pos;
        loop {
            let next = (p.0 + dx, p.1 + dy);
            if dist_sq(self.pos, next) > VISION_R_SQ {
                break;
            }
            if !self.passable(next.0, next.1) {
                break;
            }
            p = next;
        }
        dist(self.pos, p)
    }
}

impl Pathfinder for DistBug {
    fn step(&mut self) -> StepStatus {
        if self.status != StepStatus::Running {
            return self.status;
        }
        if self.pos == self.goal {
            self.status = StepStatus::Arrived;
            return self.status;
        }
        self.steps += 1;

        match self.mode {
            Mode::MotionToGoal => {
                let d = dir_to_goal(self.pos, self.goal);
                let np = neighbour(self.pos, d);
                if self.passable(np.0, np.1) {
                    self.move_to(np);
                    let new_d = dist(self.pos, self.goal);
                    if new_d < self.d_min {
                        self.d_min = new_d;
                    }
                } else {
                    self.enter_follow(d);
                }
            }
            Mode::Follow => {
                let d_leave = dist(self.pos, self.goal);
                let f = self.free_distance_to_goal();
                if d_leave - f <= self.d_min - STEP {
                    self.mode = Mode::MotionToGoal;
                    return self.status;
                }
                match self.step_wall_follow() {
                    WallStepOutcome::Moved => {}
                    WallStepOutcome::Surrounded => {
                        self.status = StepStatus::Unreachable;
                        return self.status;
                    }
                }
                self.follow_steps += 1;
                let new_d = dist(self.pos, self.goal);
                if new_d < self.d_min {
                    self.d_min = new_d;
                }
                let state = (
                    self.wf.pos,
                    self.wf.current_obstacle,
                    self.wf.obstacle_on_right,
                );
                if !self.follow_visited.insert(state) {
                    // Full circumnavigation with no qualifying leave condition.
                    self.status = StepStatus::Unreachable;
                }
            }
        }
        self.status
    }

    fn snapshot(&self) -> &Snapshot {
        &self.snap
    }

    fn summary(&self) -> String {
        let f = if self.mode == Mode::Follow {
            self.free_distance_to_goal()
        } else {
            0.0
        };
        format!(
            "pos: ({}, {})\ngoal: ({}, {})\nmode: {:?}\nhit: ({}, {})\nd_leave: {:.2}\nd_min: {:.2}\nF: {:.2}\nleave if: d_leave-F ≤ d_min-1 ({:.2} ≤ {:.2})\nobstacle: ({}, {})\nside: {}\nsensor r²≤{}\nsteps: {}\nstatus: {:?}",
            self.pos.0,
            self.pos.1,
            self.goal.0,
            self.goal.1,
            self.mode,
            self.hit_point.0,
            self.hit_point.1,
            dist(self.pos, self.goal),
            self.d_min,
            f,
            dist(self.pos, self.goal) - f,
            self.d_min - STEP,
            self.wf.current_obstacle.0,
            self.wf.current_obstacle.1,
            if self.wf.obstacle_on_right { "R" } else { "L" },
            VISION_R_SQ,
            self.steps,
            self.status,
        )
    }

    fn name(&self) -> &'static str {
        "DistBug"
    }
}

File: pkg/bench_nav/src/bench_nav/stepped/bug/distbug.py
"""DistBug — range-sensor bug, stepped."""

from __future__ import annotations

from enum import Enum
from math import sqrt
from typing import override

from bench_nav.common import INF
from bench_nav.precomputation import COST
from bench_nav.stepped.bug._common import (
    DIRS,
    VISION_R_SQ,
    WallFollowState,
    WallStepOutcome,
    dir_to_goal,
    dist_sq,
    make_passable_closures,
    neighbour,
    wall_follow_step,
)
from bench_nav.types import PrecompCtx, Stepped

_STEP = 1.0


class _Mode(Enum):
    MOTION = 0
    FOLLOW = 1


class DistBug(Stepped):
    REQUIRES = frozenset({COST})

    @override
    def __init__(self, ctx: PrecompCtx) -> None:
        self.w = ctx.w
        self.h = ctx.h
        self.cost = ctx[COST]
        self.passable, self.on_map = make_passable_closures(
            ctx.w, ctx.h, self.cost, INF
        )
        self._active_goal: int | None = None
        self._mode = _Mode.MOTION
        self._wf = WallFollowState(
            pos=(0, 0), current_obstacle=(0, 0), obstacle_on_right=True
        )
        self._d_min = 0.0
        self._follow_visited: set[tuple[tuple[int, int], tuple[int, int], bool]] = set()

    def _reset(self, p: tuple[int, int], g: tuple[int, int]) -> None:
        self._mode = _Mode.MOTION
        self._wf = WallFollowState(pos=p, current_obstacle=p, obstacle_on_right=True)
        self._d_min = sqrt(dist_sq(p, g))
        self._follow_visited.clear()

    def _free_distance_to_goal(self, cur: tuple[int, int], g: tuple[int, int]) -> float:
        d = dir_to_goal(cur, g)
        dx, dy = DIRS[d]
        p = cur
        while True:
            nxt = (p[0] + dx, p[1] + dy)
            if dist_sq(cur, nxt) > VISION_R_SQ:
                break
            if not self.passable(*nxt):
                break
            p = nxt
        return sqrt(dist_sq(cur, p))

    @override
    def step(self, pos: int, goal: int) -> int | None:
        w = self.w
        p = (pos % w, pos // w)
        g = (goal % w, goal // w)
        if goal != self._active_goal:
            self._active_goal = goal
            self._reset(p, g)
        if not self.passable(*p) or not self.passable(*g):
            return None
        if p == g:
            return pos

        if self._mode is _Mode.MOTION:
            d = dir_to_goal(p, g)
            np = neighbour(p, d)
            if self.passable(*np):
                nd = sqrt(dist_sq(np, g))
                self._d_min = min(self._d_min, nd)
                return np[1] * w + np[0]
            self._mode = _Mode.FOLLOW
            self._wf = WallFollowState(
                pos=p,
                current_obstacle=neighbour(p, d),
                obstacle_on_right=True,
            )
            self._follow_visited.clear()
            self._follow_visited.add(
                (self._wf.pos, self._wf.current_obstacle, self._wf.obstacle_on_right)
            )
            return pos
        # FOLLOW
        d_leave = sqrt(dist_sq(p, g))
        f = self._free_distance_to_goal(p, g)
        if d_leave - f <= self._d_min - _STEP:
            self._mode = _Mode.MOTION
            return pos
        self._wf.pos = p
        outcome = wall_follow_step(self._wf, self.passable, self.on_map)
        if outcome is WallStepOutcome.SURROUNDED:
            return None
        np = self._wf.pos
        nd = sqrt(dist_sq(np, g))
        self._d_min = min(self._d_min, nd)
        state = (self._wf.pos, self._wf.current_obstacle, self._wf.obstacle_on_right)
        if state in self._follow_visited:
            return None
        self._follow_visited.add(state)
        return np[1] * w + np[0]
